### PR TITLE
RES-231 Tune skyline detector

### PR DESCRIPTION
**Pull request by zuhaagha**
Ran skyline with algorithms that are commented out in https://github.com/numenta/NAB/blob/master/nab/detectors/skyline/algorithms.py. And added documentation on scoring difference and version on the scoreboard. 

Reasons why the algorithms were commented out and not included:
**grubbs** - takes over 4 hours to run
**ks_test** - incompatible. In multiprocessing/pool.py
raise self._valueTypeError: Cannot compare type 'Timestamp' with type 'float'
**is_anomalously_anomalous** - relies on redis server

### CODE REVIEW

**Comment by zuhaagha** [comment]
@subutai @BoltzmannBrain

**Comment by zuhaagha** [comment]
Addresses [Issue 109] (https://github.com/numenta/NAB/issues/109)

File: nab/detectors/skyline/skyline_detector.py [hunk 1]
```diff
@@ -11,6 +11,13 @@
 
 
 class SkylineDetector(AnomalyDetector):
+  """ Detects anomalies using Etsy Skyline's ensemble of algorithms from
+  https://github.com/etsy/skyline/blob/master/src/analyzer/algorithms.py.
+  The scoring of this detector differs slightly from Skyline. Skyline
+  relies on consensus of algorithms in the ensemble based on majority voting to
+  classify a metric as anomalous whereas this detector combines the votes of
+  all algorithms and assigns an average anomaly score.
```
Comment (BoltzmannBrain) [inline nab/detectors/skyline/skyline_detector.py]: This is a little confusing. Please describe that each algorithm in the ensemble gives a binary vote for the data record. The original implementation used a majority voting scheme to classify a record as anomalous. Here we improve the detector's performance by...

**Comment by BoltzmannBrain** [comment]
@zuhaagha did you not see score differences by trying `is_anomalously_anomalous` and `ks_test`? If they're to remain commented out, please add comments describing why.

**Comment by BoltzmannBrain** [comment]
> Addresses Issue 109

When you have a PR that upon merging will close an issue, please say "**fixes** #109", which tells GitHub to automatically close the issue when the PR is merged. If a PR is meant to address and not close an issue (i.e. more PRs are needed), then use a description other than "fixes". If that was your intent here, my bad 😉

**Comment by zuhaagha** [comment]
> If a PR is meant to address and not close an issue (i.e. more PRs are needed), then use a description other than "fixes"

Oh! Got it. It was supposed to be "fixes" then :)

**Comment by zuhaagha** [comment]
> did you not see score differences by trying is_anomalously_anomalous and ks_test?

Both of them could not run successfully. `ks_test` throws the TypeError exception mentioned in the description above. `is_anomolously_anomolous` is for second-order anomalies and runs on the history of anomalies triggered in the past, stored in redis. Its method cannot be called directly and is called through another method `run_selected_algorithm` so it cannot be integrated into our detector directly. Enabling second order anomalies is also an experimental feature in skyline as described in https://github.com/etsy/skyline/blob/master/src/settings.py.example, so its also not included in the default list of algorithms by skyline.

**Comment by BoltzmannBrain** [comment]
Did you see any score improvement by including `is_anomalously_anomalous()`?
And please add details as to why the algorithms are commented out in the algorithm script.

**Comment by zuhaagha** [comment]
> Did you see any score improvement by including is_anomalously_anomalous()?

@BoltzmannBrain `is_anomalously_anomalous()` cannot be integrated into our detector logic because it is a second-order anomaly detection method. Details are in the comment above.

**Comment by zuhaagha** [comment]
> please add details as to why the algorithms are commented out in the algorithm script.

Ok I will add that

**Comment by zuhaagha** [comment]
@BoltzmannBrain I ran `is_anomalously_anomalous` based on what you suggested yesterday. It took almost 7 hour for the run to complete and there is no change observed in the final scores.

Final score for 'skyline' detector on 'reward_low_FP_rate' profile = 27.08
Final score for 'skyline' detector on 'reward_low_FN_rate' profile = 44.48
Final score for 'skyline' detector on 'standard' profile = 35.69

**Comment by BoltzmannBrain** [comment]
@zuhaagha okay thanks. Then leave it commented out with an explanation why.

**Comment by zuhaagha** [comment]
@BoltzmannBrain Just added the explanations. Please review.

File: nab/detectors/skyline/skyline_detector.py [hunk 2]
```diff
@@ -11,6 +11,13 @@
 
 
 class SkylineDetector(AnomalyDetector):
+  """ Detects anomalies using Etsy Skyline's ensemble of algorithms from
+  https://github.com/etsy/skyline/blob/master/src/analyzer/algorithms.py.
+  Each algorithm in the ensemble gives a binary vote for the data record.
+  The original implementation used a majority voting scheme to classify a record
+  as anomalous. Here we improve the detector's performance by taking
+  average over all the votes to get an anomaly score.
```
Comment (BoltzmannBrain) [inline nab/detectors/skyline/skyline_detector.py]: "... performance by using the average of the algorithms' votes as an anomaly score."

File: nab/detectors/skyline/algorithms.py [hunk 3]
```diff
@@ -224,7 +227,8 @@ def histogram_bins(timeseries):
 
 #     return False
 
-
+# The method below is excluded because it has no affect on the final skyline
```
Comment (BoltzmannBrain) [inline nab/detectors/skyline/algorithms.py]: effect

File: nab/detectors/skyline/algorithms.py [hunk 4]
```diff
@@ -6,13 +6,13 @@
 timeseries is anomalous or not.
 
 To add an algorithm, define it here, and add its name to settings.ALGORITHMS.
+
```
Comment (BoltzmannBrain) [inline nab/detectors/skyline/algorithms.py]: Go ahead and delete the above lines.

File: nab/detectors/skyline/algorithms.py [hunk 5]
```diff
@@ -198,6 +199,8 @@ def histogram_bins(timeseries):
   return False
 
 
+# The method below is excluded because it uses floating point time whereas NAB
+# uses datetime timestamp
```
Comment (BoltzmannBrain) [inline nab/detectors/skyline/algorithms.py]: Wasn't there another reason for not including this?

**Comment by BoltzmannBrain** [comment]
@zuhaagha I left a few more comments. And please merge master into your branch to resolve the conflicts.

File: nab/detectors/skyline/algorithms.py (hunk 5 above)
Comment (zuhaagha) [inline nab/detectors/skyline/algorithms.py]: Yes there was another reason as well. I modified ks_test a little and made it work with NAB timestamp but it was slow like grubbs, so ended up terminating the run before completion. Should I change the reason to 'computationally inefficient' instead?

**Pull request merged**